src/services/WalletService.ts

```diff
@@ -133,65 +143,38 @@ export default abstract class WalletService {
+    this.adjacentUnusedAddresses = {
+      legacy: 0,
+      segwit: 0,
+      nativeSegwit: 0,
+    };
     try {
-      for (
-        let startFrom = 0;
-        startFrom < (this.getWalletMaxCall() * maxAddressPerCall) && this.subscribers.length !== 0;
-        startFrom += maxAddressPerCall
-      ) {
-        // eslint-disable-next-line no-await-in-loop
-        let addresses = await this.getAccountAddresses(maxAddressPerCall, startFrom);
-        if (addresses.length === 0) {
-          throw new Error('Error getting list of addresses - List of addresses is empty');
-        }
+      while (this.subscribers.length !== 0 && !this.areEnoughUnusedAddresses()) {
         // eslint-disable-next-line no-await-in-loop
-        addresses = await WalletService.getUnusedValue(addresses);
-        // eslint-disable-next-line no-await-in-loop
-        const balancesFound = await ApiService.getBalances(sessionId, addresses);
-        const balances = {
-          legacy: new SatoshiBig(balancesFound.legacy || 0, 'satoshi'),
-          segwit: new SatoshiBig(balancesFound.segwit || 0, 'satoshi'),
-          nativeSegwit: new SatoshiBig(balancesFound.nativeSegwit || 0, 'satoshi'),
-        };
-        if (startFrom + maxAddressPerCall >= (this.getWalletMaxCall() * maxAddressPerCall)) {
-          this.loadingBalances = false;
-        }
-        // eslint-disable-next-line no-extra-boolean-cast
-        if (!!balances) {
-          if (balances.legacy.gt(0)
-            || balances.nativeSegwit.gt(0)
-            || balances.segwit.gt(0)) {
-            balanceAccumulated = {
-              legacy: new SatoshiBig(balanceAccumulated.legacy.plus(balances.legacy), 'satoshi'),
-              segwit: new SatoshiBig(balanceAccumulated.segwit.plus(balances.segwit), 'satoshi'),
-              nativeSegwit: new SatoshiBig(balanceAccumulated.nativeSegwit.plus(balances.nativeSegwit), 'satoshi'),
-            };
-          } else {
-            const areAllAddressUnused = addresses
-              .every((walletAddressItem) => walletAddressItem.unused);
-            if (areAllAddressUnused) return;
-          }
-          this.informSubscribers(balanceAccumulated, addresses);
-        } else {
-          throw new Error('Error getting balances');
-        }
+        await this.askForBalance(sessionId, startFrom);
         const maxAmountPeginCompare = new SatoshiBig(maxAmountPegin, 'satoshi');
-        if (balanceAccumulated.legacy.gte(maxAmountPeginCompare)
-          && balanceAccumulated.segwit.gte(maxAmountPeginCompare)
-          && balanceAccumulated.nativeSegwit.gte(maxAmountPeginCompare)
+        if (this.balanceAccumulated.legacy.gte(maxAmountPeginCompare)
+          && this.balanceAccumulated.segwit.gte(maxAmountPeginCompare)
+          && this.balanceAccumulated.nativeSegwit.gte(maxAmountPeginCompare)
         ) {
           return;
         }
+        if (startFrom >= addressHardStop) {
+          throw new Error('Max address number reached');
```

Comment:
I think is better to trigger an exception because we should handle on the upper components what would happend if we don't show the complete user balance